src/core/reducer.ts

```diff
@@ -121,6 +128,77 @@ function initializeDeltalog(
   };
 }
 
+/**
+ * ExtractTransientsFromState
+ *
+ * Split out transients from the a TransientState
+ */
+function ExtractTransients(
+  transientState: TransientState
+): [State | null, TransientMetadata | null] {
+  if (!transientState) {
+    // We preserve null for the state for legacy callers, but the transient
+    // field should be undefined if not present to be consistent with the
+    // code path below.
+    return [null, undefined];
+  }
+  const { transients, ...state } = transientState;
+  return [state as State, transients as TransientMetadata];
+}
+
+/**
+ * WithError
+ *
+ * Augment a State instance with transient error information.
+ */
+function WithError<PT extends any = any>(
+  state: State,
+  errorType: ErrorType,
+  payload?: PT
+): TransientState {
+  const error = {
+    type: errorType,
+    payload,
+  };
+  return {
+    ...state,
+    transients: {
+      error,
+    },
+  };
+}
+
+/**
+ * Middleware for processing TransientState associated with the reducer
+ * returned by CreateGameReducer.
+ * This should pretty much be used everywhere you want realistic state
+ * transitions and error handling.
+ */
+export const TransientHandlingMiddleware = (store: Store) => (
```

Comment:
Could you explain what this middleware is trying to do? As above, you probably don’t want to be using the return of `store.dispatch`, but even assuming `stripTransientsResult` was the new state and not the action, wouldn’t this create a state in the following shape?

```ts
type MiddlewareReturn = State & {
  stripTransientsResult: State;
  transients: TransientMetadata;
};
```

Reply:
(I’m also not sure it’s permitted to call `dispatch` from middleware. The `dispatch` docs say [it’s an error “to call `dispatch` from inside the reducer”](https://redux.js.org/api/store#dispatchaction), but I’m not clear if there’s a distinction for middleware. Given middleware is sometimes used to handle async actions, I’m guessing maybe dispatching from middleware is supported.)